common/network-types/src/types.rs

```diff
@@ -0,0 +1,201 @@
+    /// IP address with port.
+    Ip(std::net::SocketAddr),
+}
+
+impl Display for IpOrHost {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match &self {
+            IpOrHost::Dns(host, port) => write!(f, "{host}:{port}"),
+            IpOrHost::Ip(ip) => write!(f, "{ip}"),
+        }
+    }
+}
+
+impl FromStr for IpOrHost {
+    type Err = NetworkTypeError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(addr) = std::net::SocketAddr::from_str(s) {
+            Ok(IpOrHost::Ip(addr))
+        } else {
+            s.split_once(":")
+                .ok_or(NetworkTypeError::Other("missing port delimiter".into()))
+                .and_then(|(host, port_str)| {
+                    u16::from_str(port_str)
+                        .map(|port| IpOrHost::Dns(host.to_string(), port))
+                        .map_err(|_| NetworkTypeError::Other("invalid port number".into()))
+                })
+        }
+    }
+}
+
+impl From<std::net::SocketAddr> for IpOrHost {
+    fn from(value: std::net::SocketAddr) -> Self {
+        IpOrHost::Ip(value)
+    }
+}
+
+impl std::net::ToSocketAddrs for IpOrHost {
+    type Iter = std::vec::IntoIter<std::net::SocketAddr>;
+
+    fn to_socket_addrs(&self) -> std::io::Result<Self::Iter> {
+        match &self {
+            IpOrHost::Dns(host, port) => Ok(hickory_resolver::Resolver::default()?
+                .lookup_ip(host)?
+                .into_iter()
+                .map(|ip| std::net::SocketAddr::new(ip, *port))
+                .collect::<Vec<_>>()
+                .into_iter()),
+            IpOrHost::Ip(addr) => Ok(vec![*addr].into_iter()),
+        }
+    }
+}
+
+impl IpOrHost {
+    /// Tries to resolve the DNS name and returns the first IP address found.
+    /// If this enum is already an IP address and port, it will simply return it.
+    pub fn resolve_first(self) -> Option<std::net::SocketAddr> {
+        std::net::ToSocketAddrs::to_socket_addrs(&self).ok()?.next()
+    }
+
+    /// Gets the port number.
+    pub fn port(&self) -> u16 {
+        match &self {
+            IpOrHost::Dns(_, port) => *port,
+            IpOrHost::Ip(addr) => addr.port(),
+        }
+    }
+
+    /// Gets the unresolved DNS name or IP address as string.
+    pub fn host(&self) -> String {
+        match &self {
+            IpOrHost::Dns(host, _) => host.clone(),
+            IpOrHost::Ip(addr) => addr.ip().to_string(),
+        }
+    }
+
+    /// Checks if this instance is a [DNS name](IpOrHost::Dns).
+    pub fn is_dns(&self) -> bool {
+        matches!(self, IpOrHost::Dns(..))
+    }
+
+    /// Checks if this instance is an [IP address](IpOrHost::Ip) and whether it is
+    /// an IPv4 address.
+    ///
+    /// Always returns `false` if this instance is a [DNS name](IpOrHost::Dns),
+    /// i.e.: it does not perform any DNS resolution.
+    pub fn is_ipv4(&self) -> bool {
+        matches!(self, IpOrHost::Ip(addr) if addr.is_ipv4())
+    }
```

Comment:
Isn't this redundant?

Reply:
The `IpOrHost` type is specially made to be serde serializable and transferrable. `SockAddr` always carries only IP address, never a DNS name, which it rather resolves first.

To avoid DNS name resolution at the entry, we need to be able to pass the DNS name over to the Exit node and resolve it there.